### DISCUSSION TITLE
Execute Command Script

### DISCUSSION THREAD

**Discussion by euler007** [chhoumann/quickadd#298] [Q&A]
Hi,

My query is regarding a script I am trying to create to execute a command from the command palette read the output of the plugin (i.e: Obsidian tasks plugin) and capture that into a master task file. 
I am a javascript newbie and was wondering if anyone could help me on how to execute the command in a custom script file?

**Comment by chhoumann** [comment]
Coming back to this one very late - sorry about that.

Running a command-palette command is the easy half. Two ways:

1. No code: in a Macro, add an **Obsidian Command** step and pick the Tasks command from the list.
2. From a user script: `params.app.commands.executeCommandById("obsidian-tasks-plugin:...")`. You can find any command's id by running `Object.keys(app.commands.commands)` in the developer console.

The honest catch for your use case: Obsidian commands don't return a value, so there is no "output" of a Tasks command that a script can read. Tasks query blocks render live in the note preview; they aren't text QuickAdd can capture.

Two ways to get a master task file anyway:

- Simplest: create the master file from a Template whose body contains a `tasks` (or `dataview`) query block. It renders live, so it's always current - nothing to capture.
- If you want the tasks written out as plain text, scan the vault yourself in a user script:

```js
module.exports = async (params) => {
    const lines = [];
    for (const file of params.app.vault.getMarkdownFiles()) {
        const content = await params.app.vault.cachedRead(file);
        for (const line of content.split("\n")) {
            if (line.trim().startsWith("- [ ]")) {
                lines.push(`${line.trim()} ([[${file.basename}]])`);
            }
        }
    }
    params.variables.tasks = lines.join("\n");
};
```

Then follow that step with a Capture that writes `{{VALUE:tasks}}` to your master file.

Docs on macros and script steps: https://quickadd.obsidian.guide/docs/Choices/MacroChoice
